### PR TITLE
proto: init at 0.32.2

### PR DESCRIPTION
**Pull request by nokazn**
## Description of changes

Add [moonrepo/proto](https://github.com/moonrepo/proto) newly.
Proto is a pluggable multi-language version manager.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pr/proto/package.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  doCheck = false;
+  cargoBuildFlags = [ "--bin proto" "--bin proto-shim" ];
+
+  postInstall = ''
+    # proto looks up a proto-shim executable file in $PROTO_LOOKUP_DIR
+    wrapProgram $out/bin/${pname} \
+      --set PROTO_LOOKUP_DIR $out/bin
+  '';
+
+  meta = with lib; {
```
[the remark this answers is not in the document]
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix]: Updated in https://github.com/NixOS/nixpkgs/pull/301492/commits/33faeb8faf5c8a3882c3151cbf1976c927a67ec4.

File: pkgs/by-name/pr/proto/package.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    description = "A pluggable multi-language version manager";
+    longDescription = ''
+      proto is a pluggable next-generation version manager for multiple programming languages. A unified toolchain.
+      If you use language runtimes installed by proto, you need to set `''${PROTO_HOME:-$HOME/.proto}/bin` to `PATH` in your shell startup files.
+    '';
+    homepage = "https://moonrepo.dev/proto";
+    changelog = "https://github.com/moonrepo/proto/releases/tag/v${version}";
+    license = licenses.mit;
```
[the remark this answers is not in the document]
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix]: Updated in https://github.com/NixOS/nixpkgs/commit/33faeb8faf5c8a3882c3151cbf1976c927a67ec4 same as above.

File: pkgs/by-name/pr/proto/package.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    maintainers = with maintainers; [ nokazn ];
```
[the remark this answers is not in the document]
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix]: Updated in https://github.com/NixOS/nixpkgs/commit/33faeb8faf5c8a3882c3151cbf1976c927a67ec4 same as above.

File: pkgs/by-name/pr/proto/package.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+rustPlatform.buildRustPackage rec {
+  pname = "proto";
+  version = "0.32.2";
+
+  src = fetchFromGitHub {
+    owner = "moonrepo";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-dWDHAfLg55fqPx3ZhDkA4PqoYgcJAUX43BrX43woLII=";
```
[the remark this answers is not in the document]
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix]: @msanft 
Updated to correct hashes of the revision and Cargo.lock in https://github.com/NixOS/nixpkgs/pull/301492/commits/33faeb8faf5c8a3882c3151cbf1976c927a67ec4.
Sorry, it was of a invalid revision. I don't know why but didn't cause mismatch errors in my side at that time. :(

File: pkgs/by-name/pr/proto/package.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    # proto looks up a proto-shim executable file in $PROTO_LOOKUP_DIR
+    wrapProgram $out/bin/${pname} \
+      --set PROTO_LOOKUP_DIR $out/bin
+  '';
+
+  meta = {
+    description = "A pluggable multi-language version manager";
+    longDescription = ''
+      proto is a pluggable next-generation version manager for multiple programming languages. A unified toolchain.
+      If you use language runtimes installed by proto, you need to set `''${PROTO_HOME:-$HOME/.proto}/bin` to `PATH` in your shell startup files.
```
Comment (msanft) [inline pkgs/by-name/pr/proto/package.nix:44]: I'm unsure if this shouldn't rather be a comment in the package. If this is mentioned somewhere in proto's own documentation, we should drop it here. If not, and if it's not NixOS-specific, I'm fine with having it.

File: pkgs/by-name/pr/proto/package.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    '';
+    homepage = "https://moonrepo.dev/proto";
+    changelog = "https://github.com/moonrepo/proto/releases/tag/v${version}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ nokazn ];
+    mainProgram = "proto";
+  };
```
Comment (msanft) [inline pkgs/by-name/pr/proto/package.nix:51]:
```suggestion
    platforms = lib.platforms.linux ++ lib.platforms.darwin;
  };
```

**Review by msanft — CHANGES_REQUESTED** [review]
Tested the binary, which works on `x86_64-linux`.

**Comment by msanft** [comment]
Result of `nixpkgs-review pr 301492` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>proto</li>
  </ul>
</details>

File: pkgs/by-name/pr/proto/package.nix (hunk 6 above)
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix:51]: Added in https://github.com/NixOS/nixpkgs/pull/301492/commits/b42ac809231651028ac50ffb38d41857348b9ee1. Thank you.

File: pkgs/by-name/pr/proto/package.nix (hunk 5 above)
Comment (nokazn) [reply pkgs/by-name/pr/proto/package.nix]: > If this is mentioned somewhere in proto's own documentation, we should drop it here.

Indeed this is documented in https://moonrepo.dev/docs/proto, so dropped the last line. https://github.com/NixOS/nixpkgs/pull/301492/commits/b42ac809231651028ac50ffb38d41857348b9ee1

**Review by msanft — APPROVED** [review]
Thanks!

**Comment by lolbinarycat** [comment]
i am a bit confused with the name "proto", as it doesn't seem to have anything to do with prototyping or protocols.

**Comment by msanft** [comment]
> i am a bit confused with the name "proto", as it doesn't seem to have anything to do with prototyping or protocols.

I was a little confused as well about their naming choice, but they've got some [reference](https://moonrepo.dev/docs/proto/faq#where-did-the-name-proto-come-from) on it

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**